File: gym/blade/units/Facility.py
import json
from typing import List, Optional
from blade.units.Weapon import Weapon
from blade.utils.colors import convert_color_name_to_side_color, SIDE_COLOR
from blade.units.combat_stats import (
    apply_damage,
    get_health_fraction,
    resolve_current_hp,
    resolve_defense,
    resolve_max_hp,
)


class Facility:
    def __init__(
        self,
        id: str,
        name: str,
        side_id: str,
        class_name: str,
        latitude: float = 0.0,
        longitude: float = 0.0,
        altitude: float = 0.0,  # FT ASL -- currently default
        range: float = 250.0,
        side_color: str | SIDE_COLOR | None = None,
        weapons: Optional[List[Weapon]] = None,
        max_hp: float | None = None,
        current_hp: float | None = None,
        defense: float | None = None,
    ):
        self.id = id
        self.name = name
        self.side_id = side_id
        self.class_name = class_name
        self.latitude = latitude
        self.longitude = longitude
        self.altitude = altitude
        self.range = range
        self.side_color = convert_color_name_to_side_color(side_color)
        self.weapons = weapons if weapons is not None else []
        self.max_hp = resolve_max_hp("facility", max_hp)
        self.current_hp = resolve_current_hp(current_hp, self.max_hp)
        self.defense = resolve_defense("facility", defense)

    def get_total_weapon_quantity(self) -> int:
        return sum([weapon.current_quantity for weapon in self.weapons])

    def get_weapon_with_highest_engagement_range(self) -> Weapon | None:
        if len(self.weapons) == 0:
            return None
        return max(self.weapons, key=lambda weapon: weapon.get_engagement_range())

    def get_detection_range(self) -> float:
        return self.range

    def get_health_fraction(self) -> float:
        return get_health_fraction(self.current_hp, self.max_hp)

    def apply_damage(self, raw_attack: float) -> float:
        return apply_damage(self, raw_attack)

    def is_destroyed(self) -> bool:
        return self.current_hp <= 0

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name,
            "side_id": str(self.side_id),
            "class_name": self.class_name,
            "latitude": self.latitude,
            "longitude": self.longitude,
            "altitude": self.altitude,
            "range": self.range,
            "side_color": (
                self.side_color.value
                if isinstance(self.side_color, SIDE_COLOR)
                else self.side_color
            ),
            "weapons": [weapon.to_dict() for weapon in self.weapons],
            "max_hp": self.max_hp,
            "current_hp": self.current_hp,
            "defense": self.defense,
        }

File: client/src/game/units/Facility.ts
import Weapon from "@/game/units/Weapon";
import { convertColorNameToSideColor, SIDE_COLOR } from "@/utils/colors";
import { getFacilityDetectionArcDegrees } from "@/game/db/facilityThreatProfiles";
import {
  computeDamage,
  getHealthFraction,
  resolveCurrentHp,
  resolveDefense,
  resolveMaxHp,
} from "@/game/units/combatStats";

interface IFacility {
  id: string;
  name: string;
  sideId: string;
  className: string;
  latitude: number;
  longitude: number;
  altitude: number;
  range: number;
  heading?: number;
  speed?: number;
  route?: number[][];
  sideColor?: string | SIDE_COLOR;
  weapons?: Weapon[];
  detectionArcDegrees?: number;
  maxHp?: number;
  currentHp?: number;
  defense?: number;
}

export default class Facility {
  id: string;
  name: string;
  sideId: string;
  className: string;
  latitude: number = 0.0;
  longitude: number = 0.0;
  altitude: number = 0.0; // FT ASL -- currently default -- need to reference from database
  range: number = 250; // NM -- currently default -- need to reference from database
  heading: number = 0.0;
  speed: number = 0.0; // KTS
  route: number[][] = [];
  sideColor: SIDE_COLOR;
  weapons: Weapon[] = [];
  detectionArcDegrees: number;
  maxHp: number;
  currentHp: number;
  defense: number;

  constructor(parameters: IFacility) {
    this.id = parameters.id;
    this.name = parameters.name;
    this.sideId = parameters.sideId;
    this.className = parameters.className;
    this.latitude = parameters.latitude;
    this.longitude = parameters.longitude;
    this.altitude = parameters.altitude;
    this.range = parameters.range;
    this.heading = parameters.heading ?? 0.0;
    this.speed = parameters.speed ?? 0.0;
    this.route = parameters.route ?? [];
    this.sideColor = convertColorNameToSideColor(parameters.sideColor);
    this.weapons = parameters.weapons ?? [];
    this.detectionArcDegrees =
      parameters.detectionArcDegrees ??
      getFacilityDetectionArcDegrees(parameters.className);
    this.maxHp = resolveMaxHp("facility", parameters.maxHp);
    this.currentHp = resolveCurrentHp(parameters.currentHp, this.maxHp);
    this.defense = resolveDefense("facility", parameters.defense);
  }

  getTotalWeaponQuantity(): number {
    let sum = 0;
    this.weapons.forEach((weapon) => {
      sum += weapon.currentQuantity;
    });
    return sum;
  }

  getWeaponWithHighestEngagementRange(): Weapon | undefined {
    if (this.weapons.length === 0) return;
    return this.weapons.reduce((a, b) =>
      a.getEngagementRange() > b.getEngagementRange() ? a : b
    );
  }

  getWeaponEngagementRange(): number {
    if (this.weapons.length === 0) return 0;
    return (
      this.getWeaponWithHighestEngagementRange()?.getEngagementRange() ?? 0
    );
  }

  getDetectionRange(): number {
    return this.range;
  }

  getDetectionArcDegrees(): number {
    return this.detectionArcDegrees;
  }

  getDetectionHeading(): number {
    return this.heading;
  }

  getHealthFraction(): number {
    return getHealthFraction(this.currentHp, this.maxHp);
  }

  applyDamage(rawAttackPower: number): number {
    const damage = computeDamage(rawAttackPower, this.defense);
    this.currentHp = Math.max(this.currentHp - damage, 0);
    return damage;
  }

  isDestroyed(): boolean {
    return this.currentHp <= 0;
  }
}
